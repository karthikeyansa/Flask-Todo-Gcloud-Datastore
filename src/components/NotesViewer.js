import React, {useEffect, useState} from "react";
import {API} from "../Apiservice";


window.completedTasks = 0;
window.ongoingTasks = 0;
window.note = {};

function NotesViewer(props){

    const [note] = useState(props.note);
    const [noteContent, setNoteContent] = useState("");
    const [newTask, setNewtask] = useState("");


    useEffect(()=>{
        countKeeper(window.completedTasks, window.ongoingTasks);
    },[])

    const countKeeper = (completedTasks, ongoingTasks) =>{
        note.tasks.forEach(task => {
            if(task.status === false){
                ongoingTasks += 1;
            }
            else if(task.status === true){
                completedTasks += 1;
            }
            window.ongoingTasks = ongoingTasks; 
            window.completedTasks = completedTasks;
            window.note = props.note;
        })
        let ongoingTasksCount = document.getElementById("incompleteTaskCount");
        if(ongoingTasksCount) ongoingTasksCount.innerText = `‚ùå Ongoing (${window.ongoingTasks})`;
        let completedTasksCount = document.getElementById("completeTaskCount");
        if(completedTasksCount) completedTasksCount.innerHTML = `<span style={{color:"lightgreen"}} >‚úî </span>Completed (${window.completedTasks})`;
    } 

    const expandTextarea = (event) =>{
        event.target.style.height="100px";
    }

    const noteContentarea = async(event) =>{
        let alerts = document.getElementById("alerts");
        if(noteContent !== ""){
            try{
                let response = await API.UpdateNoteandTaskcontent({note_id: note.id, content: noteContent});
                if(response.success === true){
                    window.note = response.result;
                    alerts.style.display = "block";
                    alerts.style.backgroundColor = "green";
                    alerts.innerText = "Note story updated";
                    setTimeout(()=>{
                        alerts.style.display = "none"
                    }, 1000)
                    setNoteContent("");
                }
            }
            catch(error){
                console.log(error);
            }
        }
        else if(newTask !== ""){
            try{
                let response = await API.UpdateNoteandTaskcontent({note_id: note.id, description: newTask});
                if(response.success === true){
                    window.note = response.result;
                    response.result.tasks.filter((task)=>{
                    if(task.description === newTask){
                        let incompleteTasks = document.getElementById("incompleteTasks");
                        
                        let taskStatusButton = document.createElement("span");
                        taskStatusButton.setAttribute("class", "modalIcon");
                        taskStatusButton.setAttribute("id", `span${task.task_id}`);
                        taskStatusButton.addEventListener("click", function(event){
                            checkStatus(note.id, task.task_id, event.target.innerText)
                        })
                        taskStatusButton.innerText = "‚ùå";
                        
                        let taskDescription = document.createElement("textarea");
                        taskDescription.setAttribute("class", "taskTextarea");
                        taskDescription.setAttribute("id", `descript${task.task_id}`);
                        taskDescription.innerHTML = newTask;
                        taskDescription.addEventListener("blur", function(){
                            noteTaskarea(note.id, task.task_id)
                        })  
                        taskDescription.addEventListener("focusin", function(){
                            taskDescription.style.height = "100px";
                            incompleteTasks.scrollTop = incompleteTasks.scrollHeight;
                        })
                                                    
                        let taskDeleteButton = document.createElement("span");
                        taskDeleteButton.setAttribute("id", `trash${task.task_id}`);
                        taskDeleteButton.setAttribute("class", "modalIcon");
                        taskDeleteButton.innerText = "üóëÔ∏è";
                        taskDeleteButton.addEventListener("click", function(){
                            taskRemover(note.id, task.task_id)
                        })
    
                        incompleteTasks.appendChild(taskStatusButton);
                        incompleteTasks.appendChild(taskDescription);
                        incompleteTasks.appendChild(taskDeleteButton);
                        incompleteTasks.scrollTop = incompleteTasks.scrollHeight;
    
                        let ongoingTasksCount = document.getElementById("incompleteTaskCount");
                        window.ongoingTasks = window.ongoingTasks + 1;
                        if(ongoingTasksCount) ongoingTasksCount.innerText = `‚ùå Ongoing (${window.ongoingTasks})`;
                    }})
                    setNewtask("");
                    let alerts = document.getElementById("alerts");
                    alerts.style.display = "block";
                    alerts.style.backgroundColor = "green";
                    alerts.innerText = "New task added.";
                    setTimeout(()=>{
                        alerts.style.display = "none"
                    }, 1000)
                }
            }
            catch(error){
                console.log(error);
            }
            
        }
    }

    const closeModal = () => {
        window.completedTasks = 0;
        window.ongoingTasks = 0;
        props.ModalClosed(window.note);

    }
    return(
        <React.Fragment>
            <br />
            <div className="alerts" id="alerts"></div>
            <button onClick={closeModal} className="modalCloseButton danger-btn">Close</button>
            <br />
            <div className="noteViewerGrid">
                <textarea className="noteContentTextarea" id="NoteContent" 
                          onBlur={noteContentarea} onChange={(e)=>setNoteContent(e.target.value)} defaultValue={note.content}></textarea>
                <div className="ongoingTasks"  id="incompleteTaskCount"></div>
                <div className="categoryTasks" id="incompleteTasks">
                    {note.tasks && note.tasks.filter((task)=>task.status === false).map((task)=>{
                        return(
                        <React.Fragment key={task.task_id}>
                            <span id={`span${task.task_id}`} style={{textAlign: "center"}} className="modalIcon" 
                                    onClick={(e) => {checkStatus(note.id, task.task_id, e.target.innerText)}}>‚ùå</span>
                            <textarea className="taskTextarea" 
                                        onBlur={()=>noteTaskarea(note.id, task.task_id)} 
                                        onSelect={expandTextarea} id={`descript${task.task_id}`} defaultValue={task.description}></textarea>
                            <span id={`trash${task.task_id}`} className="modalIcon" 
                                    onClick={()=>taskRemover(note.id, task.task_id)}>üóëÔ∏è</span>
                        </React.Fragment>
                        )
                    })}
                </div>
                <br/>
                <div className="newTaskTextarea" style={{height: "55px"}}>
                    <textarea className="newtexareaPadding" style={{height: "50px"}} 
                              placeholder="Add new task here." onChange={(e)=>{setNewtask(e.target.value)}} value={newTask}></textarea>
                    <span>&nbsp;</span>
                    <div className="newTaskButtonGrid">
                        <button className="newNoteButtons success-btn" onClick={noteContentarea}>Add</button>
                        <span style={{height:"0px"}}>&nbsp;</span>
                        <button className="newNoteButtons danger-btn" onClick={(e)=>setNewtask("")}>Cancel</button>
                    </div>
                </div>
                <br />
                <div className="completedTasks" id="completeTaskCount"></div>
                <div className="categoryTasks" id="completeTasks">
                    {note.tasks && note.tasks.filter((task)=>task.status === true).map((task)=>{
                        return(
                        <React.Fragment key={task.task_id}>
                            <span id={`span${task.task_id}`} className="modalIcon"
                                    onClick={(e) => {checkStatus(note.id, task.task_id, e.target.innerText)}}>‚úî</span>
                            <textarea className="taskTextarea" 
                                        onBlur={()=>noteTaskarea(note.id, task.task_id)} 
                                        onSelect={expandTextarea} id={`descript${task.task_id}`} defaultValue={task.description}></textarea>
                            <span id={`trash${task.task_id}`} className="modalIcon" 
                                    onClick={()=>taskRemover(note.id, task.task_id)}>üóëÔ∏è</span>
                        </React.Fragment>
                        )
                    })}
                </div>
            </div>
        </React.Fragment>
    )
}

async function checkStatus(note_id, task_id, status){
    var response = {};
    try{
        if(status === '‚úî'){
            response = await API.StatusUpdate({note_id: note_id, task_id: task_id, status: false})
        }
        else if(status === '‚ùå'){
            response = await API.StatusUpdate({note_id: note_id, task_id: task_id, status: true})
        }
    }
    catch(error){
        console.log(error);
    }
    if(response.success === true){
        window.note = response.result;
        let taskStatusButton = document.getElementById(`span${task_id}`);
        let taskdescript = document.getElementById(`descript${task_id}`);
        let trashButton = document.getElementById(`trash${task_id}`);
        taskStatusButton.remove();
        taskdescript.remove();
        trashButton.remove();
        if (taskStatusButton.innerText === "‚ùå") {
            taskStatusButton.innerText = "‚úî";
            let completedTasks = document.getElementById("completeTasks");
            completedTasks.appendChild(taskStatusButton);
            completedTasks.appendChild(taskdescript);
            completedTasks.appendChild(trashButton);
            completedTasks.scrollTop = completedTasks.scrollHeight;
            let completedTasksCount = document.getElementById("completeTaskCount");
            window.completedTasks = window.completedTasks + 1;
            if(completedTasksCount) completedTasksCount.innerHTML = `<span style={{color:"lightgreen"}} >‚úî </span>Completed (${window.completedTasks})`;
            // deducte by one
            let ongoingTasksCount = document.getElementById("incompleteTaskCount");
            window.ongoingTasks = window.ongoingTasks - 1;
            if(ongoingTasksCount) ongoingTasksCount.innerText = `‚ùå Ongoing (${window.ongoingTasks})`;
        
        }
        else if (taskStatusButton.innerText === "‚úî") {
            taskStatusButton.innerText = "‚ùå";
            let ongoingTasks = document.getElementById("incompleteTasks")
            ongoingTasks.appendChild(taskStatusButton);
            ongoingTasks.appendChild(taskdescript);
            ongoingTasks.appendChild(trashButton);
            ongoingTasks.scrollTop = ongoingTasks.scrollHeight;
            let ongoingTasksCount = document.getElementById("incompleteTaskCount");
            window.ongoingTasks = window.ongoingTasks + 1;
            if(ongoingTasksCount) ongoingTasksCount.innerText = `‚ùå Ongoing (${window.ongoingTasks})`;
            // deducte by one
            let completedTasksCount = document.getElementById("completeTaskCount");
            window.completedTasks = window.completedTasks - 1;
            if(completedTasksCount) completedTasksCount.innerHTML = `<span style={{color:"lightgreen"}} >‚úî </span>Completed (${window.completedTasks})`;
           
        }    
    }
}

async function taskRemover(note_id, task_id){
    let taskStatusButton = document.getElementById(`span${task_id}`);
    let taskdescript = document.getElementById(`descript${task_id}`);
    let taskDeleteButton = document.getElementById(`trash${task_id}`);
    let alerts = document.getElementById("alerts");
    let ongoingTasksCount = document.getElementById("incompleteTaskCount");
    let completedTasksCount = document.getElementById("completeTaskCount");
    try{
        let response = await API.RemoveTask({note_id: note_id, task_id: task_id});
        if(response.success === true){
            window.note = response.result;
            taskStatusButton.remove();
            taskdescript.remove();
            taskDeleteButton.remove();
            alerts.style.display = "block";
            alerts.style.backgroundColor = "green";
            alerts.innerText = "Task removed successfully";
            setTimeout(()=>{
                alerts.style.display = "none"
            }, 1000)
            if(taskStatusButton.innerText === "‚ùå"){
                window.ongoingTasks = window.ongoingTasks - 1;
                if(ongoingTasksCount) ongoingTasksCount.innerText = `‚ùå Ongoing (${window.ongoingTasks})`;
            }
            else if(taskStatusButton.innerText === "‚úî"){
                window.completedTasks = window.completedTasks - 1;
                if(completedTasksCount) completedTasksCount.innerHTML = `<span style={{color:"lightgreen"}} >‚úî </span>Completed (${window.completedTasks})`;
            }
        }
    }
    catch(error){
        console.log(error);
    }

}

async function noteTaskarea(note_id=0, task_id=0){
    let taskarea = document.getElementById(`descript${task_id}`);
    let alerts = document.getElementById("alerts");
    taskarea.style.height = "25px";
    if(taskarea.value !== ""){
        try{
            let response = await API.UpdateTaskDescription({note_id: note_id, task_id: task_id, description: taskarea.value});
            if(response.success === true){
                window.note = response.result;
                alerts.style.display = "block";
                alerts.style.backgroundColor = "green";
                alerts.innerText = "Task updated";
                setTimeout(()=>{
                    alerts.style.display = "none"
                }, 1000)
            }
        }
        catch(error){
            console.log(error);
        }
    }
}

export default NotesViewer;