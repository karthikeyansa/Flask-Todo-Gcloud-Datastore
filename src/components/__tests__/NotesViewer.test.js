import React from 'react';
import { mount, ReactWrapper, shallow } from 'enzyme';

import NotesViewer from '../NotesViewer';
import renderer from 'react-test-renderer';

const simulateOnChangeInput = (wrapper, inputSelector, newValue) =>{
  wrapper.find(inputSelector).first().simulate('change', {
      target: { value: newValue},
  })
  return wrapper.find(inputSelector).first();
}

// const simulateClick = (wrapper, buttonSelector, eventAction) => {
//   wrapper.find(buttonSelector).first().simulate(eventAction)
//   return wrapper.find(buttonSelector).first();
// }

describe('NotesViewer', () => {
  it('should render', () => {
    const note = {id: 1, content: "Hello world", 
                  tasks: [{task_id: 1, description: "this is task 1", status: true}, 
                          {task_id: 2, description: "this is task 2", status: false}]}
    const wrapper = shallow(<NotesViewer note={note} />)
    expect(wrapper.exists()).toBeTruthy();
  })
})


describe('NotesViewer', () => {
  it('Alerts Div', () => {
    const note = {id: 1, content: "Hello world", 
                  tasks: [{task_id: 1, description: "this is task 1", status: true}, 
                          {task_id: 2, description: "this is task 2", status: false}]}
      const wrapper = shallow(<NotesViewer note={note} />);
      expect(wrapper.find("#alerts").text()).toBe("");
      expect(wrapper.find("button").find(".modalCloseButton").text()).toBe("Close");
  })
})

describe('NotesViewer', () => {
  it('Viewer Grid', () => {
    const note = {id: 1, content: "Hello world", 
                  tasks: [{task_id: 1, description: "this is task 1", status: true}, 
                          {task_id: 2, description: "this is task 2", status: false}]}
      const wrapper = shallow(<NotesViewer note={note} />);
      expect(wrapper.find(".noteViewerGrid")).toBeTruthy();
      expect(wrapper.find(".noteViewerGrid").find("#NoteContent").text()).toBe("");
      expect(wrapper.find(".noteViewerGrid").find("#incompleteTaskCount").text()).toBe("");
      expect(wrapper.find(".noteViewerGrid").find("#incompleteTasks").text()).toBe("❌🗑️");
      expect(wrapper.find(".noteViewerGrid").find("#completeTaskCount").text()).toBe("");
      expect(wrapper.find(".noteViewerGrid").find("#completeTasks").text()).toBe("✔🗑️");
  })
})

describe('NotesViewer', () => {
  const note = {id: 1, content: "Hello world", 
                  tasks: [{task_id: 1, description: "this is task 1", status: true}, 
                          {task_id: 2, description: "this is task 2", status: false}]};
  it('New Task Textarea', () => {
      const wrapper = shallow(<NotesViewer note={note} />);
      expect(wrapper.find(".noteViewerGrid")).toBeTruthy();
      expect(wrapper.find(".noteViewerGrid #newTaskTextarea")).toBeTruthy();
  })
})

describe('NotesViewer', () => {
  const holder = document.createElement('div');
  document.body.appendChild(holder);
  
  const note = {id: 1, content: "Hello world", 
                  tasks: [{task_id: 1, description: "this is task 1", status: true}, 
                          {task_id: 2, description: "this is task 2", status: false}]};

  it('Note renderning and close actions block', () =>{
    const wrapper = mount(<NotesViewer note={note} />, { attachTo: holder });
    const closeModelButtton = wrapper.find("button.modalCloseButton.danger-btn").first();
    expect(closeModelButtton.text()).toBe("Close");

    wrapper.unmount();

  })
})

describe('NoteViewer', () =>{
  const holder = document.createElement('div');
  document.body.appendChild(holder);

  const mockFetch = Promise.resolve({ json: () => Promise.resolve(
    {success: true, result:[{id: 1, content: "Hello world", 
                            tasks: [{task_id: 1, description: "this is task 1", status: true}, 
                                    {task_id: 2, description: "this is task 2", status: false}]},
                            {id: 2, content: "Hello mars", 
                            tasks: [{task_id: 1, description: "this is task 3", status: true}, 
                                    {task_id: 2, description: "this is task 4", status: true}]}]
    })
  });
  global.fetch = jest.fn().mockImplementation(() => mockFetch);

  const minprops = {
    note : {id: 1, content: "Hello world", 
                    tasks: [{task_id: 1, description: "this is task 1", status: true}, 
                            {task_id: 2, description: "this is task 2", status: false}]}
  }

  beforeEach(()=>{
    jest.useFakeTimers();
  })

  afterEach(()=>{
    jest.runAllTimers();
  })

  it('Note renderer update Note Content and task description', () => {

    const wrapper = mount(<NotesViewer {...minprops}/>, { attachTo: holder });
    
    simulateOnChangeInput(wrapper, "div.noteViewerGrid textarea#NoteContent", "hello mars");
    const checkStatusButton = wrapper.find("div.noteViewerGrid div#incompleteTasks span#span2").first();
    checkStatusButton.simulate('click');
    const taskDescriptionUpdate = wrapper.find("div.noteViewerGrid div#incompleteTasks textarea.taskTextarea").first();
    taskDescriptionUpdate.simulate('blur');
    const trashButton = wrapper.find("div.noteViewerGrid div#incompleteTasks span#trash2").first();
    trashButton.simulate('click');
    //Error Handling 
    wrapper.find("div.noteViewerGrid").find("div.newTaskTextarea").find("textarea.newtexareaPadding").simulate("change", {target: { value: 'this is new task'}});
    wrapper.find("div.noteViewerGrid").find("div.newTaskTextarea").find("textarea.newtexareaPadding").first();
    expect(wrapper.find("div.noteViewerGrid").find("div.newTaskTextarea").find("textarea.newtexareaPadding").text()).toBe("this is new task");

    wrapper.find("div.noteViewerGrid").find("div.newTaskTextarea").find("div.newTaskButtonGrid button.newNoteButtons.success-btn").simulate('click')
    wrapper.find("div.noteViewerGrid").find("div.newTaskTextarea").find("div.newTaskButtonGrid button.newNoteButtons.danger-btn").simulate('click')

    const checkStatusButtonCompleted = wrapper.find("div.noteViewerGrid div#completeTasks span#span1").first();
    checkStatusButtonCompleted.simulate('click');
    const taskDescriptionUpdateCompleted = wrapper.find("div.noteViewerGrid div#completeTasks textarea.taskTextarea").first();
    taskDescriptionUpdateCompleted.simulate('blur');
    const trashButtonCompleted = wrapper.find("div.noteViewerGrid div#completeTasks span#trash1").first();
    trashButtonCompleted.simulate('click');

    wrapper.unmount();
  })
})

describe("NoteViewer", () => {
  let wrapper;
  beforeEach(()=>{
    const mockFetch = Promise.resolve({ json: () => Promise.resolve(
    {success: true, result:{
      id: 1, content: "Hello world", 
      tasks: [{task_id: 1, description: "this is task 1", status: true}, 
              {task_id: 2, description: "this is task 2", status: false}]
    }
    })
    });

    global.fetch = jest.fn().mockImplementation(() => mockFetch);
  
    const minprops = {
      note : {id: 1, content: "Hello world", 
                      tasks: [{task_id: 1, description: "this is task 1", status: true}, 
                              {task_id: 2, description: "this is task 2", status: false}]}
    }

    wrapper = mount(<NotesViewer {...minprops} />);
  })

  it("Note rendenering new task description", () => {
    expect(wrapper.find("div.noteViewerGrid").find("div.newTaskTextarea").find("textarea.newtexareaPadding").exists()).toBeTruthy();
    wrapper.find("div.noteViewerGrid").find("div.newTaskTextarea").find("textarea.newtexareaPadding").simulate("change", {target: { value: 'this is task one'}});
    wrapper.find("div.noteViewerGrid").find("div.newTaskTextarea").find("textarea.newtexareaPadding").first();
    expect(wrapper.find("div.noteViewerGrid").find("div.newTaskTextarea").find("textarea.newtexareaPadding").text()).toBe("this is task one");

    wrapper.find("div.noteViewerGrid").find("div.newTaskTextarea").find("div.newTaskButtonGrid button.newNoteButtons.success-btn").simulate('click')
    wrapper.find("div.noteViewerGrid").find("div.newTaskTextarea").find("div.newTaskButtonGrid button.newNoteButtons.danger-btn").simulate('click')
  })

  afterEach(()=>{
    jest.clearAllMocks();
    wrapper.unmount();
  })
})


describe('NotesViewer', () => {
  it('should render without crashing as per in snapshot', () => {
    const tree = renderer.create(<NotesViewer note={note}/>).toJSON();
    expect(tree).toMatchSnapshot();
  })

})