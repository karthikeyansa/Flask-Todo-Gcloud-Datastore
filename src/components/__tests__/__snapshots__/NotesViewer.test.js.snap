// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotesViewer should render without crashing as per in snapshot 1`] = `
Array [
  <br />,
  <div
    className="alerts"
    id="alerts"
  />,
  <button
    className="modalCloseButton danger-btn"
    onClick={[Function]}
  >
    Close
  </button>,
  <br />,
  <div
    className="noteViewerGrid"
  >
    <textarea
      className="noteContentTextarea"
      defaultValue="Hello world"
      id="NoteContent"
      onBlur={[Function]}
      onChange={[Function]}
    />
    <div
      className="ongoingTasks"
      id="incompleteTaskCount"
    />
    <div
      className="categoryTasks"
      id="incompleteTasks"
    >
      <span
        className="modalIcon"
        id="span2"
        onClick={[Function]}
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        ❌
      </span>
      <textarea
        className="taskTextarea"
        defaultValue="this is task 2"
        id="descript2"
        onBlur={[Function]}
        onSelect={[Function]}
      />
      <span
        className="modalIcon"
        id="trash2"
        onClick={[Function]}
      >
        🗑️
      </span>
    </div>
    <br />
    <div
      className="newTaskTextarea"
      style={
        Object {
          "height": "55px",
        }
      }
    >
      <textarea
        className="newtexareaPadding"
        onChange={[Function]}
        placeholder="Add new task here."
        style={
          Object {
            "height": "50px",
          }
        }
        value=""
      />
      <span>
         
      </span>
      <div
        className="newTaskButtonGrid"
      >
        <button
          className="newNoteButtons success-btn"
          onClick={[Function]}
        >
          Add
        </button>
        <span
          style={
            Object {
              "height": "0px",
            }
          }
        >
           
        </span>
        <button
          className="newNoteButtons danger-btn"
          onClick={[Function]}
        >
          Cancel
        </button>
      </div>
    </div>
    <br />
    <div
      className="completedTasks"
      id="completeTaskCount"
    />
    <div
      className="categoryTasks"
      id="completeTasks"
    >
      <span
        className="modalIcon"
        id="span1"
        onClick={[Function]}
      >
        ✔
      </span>
      <textarea
        className="taskTextarea"
        defaultValue="this is task 1"
        id="descript1"
        onBlur={[Function]}
        onSelect={[Function]}
      />
      <span
        className="modalIcon"
        id="trash1"
        onClick={[Function]}
      >
        🗑️
      </span>
    </div>
  </div>,
]
`;
